packages/SwingSet/src/kernel/loadVat.js

```diff
@@ -0,0 +1,278 @@
+   *        `setup()` API, which bypasses the imposition of LiveSlots but
+   *        requires the vat implementation to enforce the vat invariants
+   *        manually.  If false, the vat will be constructed using the
+   *        `buildRootObject()` API, which uses LiveSlots to enforce the vat
+   *        invariants automatically.  Defaults to false.
+   *
+   *    'enablePipelining' If true, permits the kernel to pipeline messages to
+   *        promises for which the vat is the decider directly to the vat
+   *        without waiting for the promises to be resolved.  If false, such
+   *        messages will be queued inside the kernel.  Defaults to false.
+   *
+   * @param {boolean} notifyNewVat  If true, the success or failure of the
+   *    operation will be reported in a message to the admin vat, citing the
+   *    vatID
+   * @param {boolean} isDynamic  If true, the vat being created is a dynamic vat;
+   *    if false, it's a static vat (these have differences in their allowed
+   *    options and some of their option defaults).
+   *
+   * @returns {string} the vatID for a newly created vat.
+   */
+  function create(vatID, source, options, notifyNewVat, isDynamic) {
+    assert(source.bundle || source.bundleName, 'broken source');
+    const vatSourceBundle =
+      source.bundle || kernelKeeper.getBundle(source.bundleName);
+    if (!vatSourceBundle) {
+      throw Error(`Bundle ${source.bundleName} not found`);
+    }
+    // TODO: maybe hash the bundle object somehow for the description
+    const description = source.bundle
+      ? '(source bundle)'
+      : `from: ${source.bundleName}`;
+
+    assertKnownOptions(
+      options,
+      isDynamic ? allowedDynamicOptions : allowedStaticOptions,
+    );
+    const {
+      metered = isDynamic,
+      vatParameters = {},
+      enableSetup = false,
+      enablePipelining = false,
+    } = options;
+    let terminated = false;
+
+    function notifyTermination(error) {
+      if (terminated) {
+        return;
+      }
+      terminated = true;
+      const vatAdminVatId = vatNameToID('vatAdmin');
+      const vatAdminRootObjectSlot = makeVatRootObjectSlot();
+
+      const args = {
+        body: JSON.stringify([
+          vatID,
+          error
+            ? { '@qclass': 'error', name: error.name, message: error.message }
+            : { '@qclass': 'undefined' },
+        ]),
+        slots: [],
+      };
+
+      queueToExport(
+        vatAdminVatId,
+        vatAdminRootObjectSlot,
+        'vatTerminated',
+        args,
+        'logFailure',
+      );
+    }
+
+    async function build() {
+      if (typeof vatSourceBundle !== 'object') {
+        throw Error(`vat creation requires a bundle, not a plain string`);
+      }
+
+      kernelSlog.addVat(vatID, true, description);
+      const managerOptions = {
+        bundle: vatSourceBundle,
+        metered,
+        enableSetup,
+        enablePipelining,
+        enableInternalMetering: !isDynamic,
```

Comment:
we lost a TODO comment about #1343 that was in kernel.js inside `addGenesisVat`, please copy it in here to remind us to remove `enableInternalMetering` eventually:

```
    // TODO: We need to support within-vat metering (for the Spawner) until
    // #1343 is fixed, after which we can remove
    // managerOptions.enableInternalMetering . For now, it needs to be
    // enabled for our internal unit test (which could easily add this to its
    // config object) and for the spawner vat (not so easy). To avoid deeper
    // changes, we enable it for *all* static vats here. Once #1343 is fixed,
    // remove this addition and all support for internal metering.
```